### ISSUE TITLE
SQL Server transport does not allow to easily override the catalog for all queues/endpoints

### ISSUE THREAD

**Issue by SzymonPobiega** [Particular/NServiceBus.SqlServer#751]
SQL Server transport uses the catalog from the connection string (`initial catalog` property) when accessing DB objects. This can be overridden on per-endpoint or per-queue basis using `UseCatalogPerQueue` nad `UseCatalogPerEndpoint`. These APIs allow users to set up a _catalog-per-endpoint_ topology in which each endpoint uses a separate catalog for data (business + queues).

It does not allow, however, to set up a _queues-in-shared-db_ topology easily because all destination queues have to be configured via these APIs. Event worse, if a new endpoint is added, all other endpoints need to be re-configured and re-deployed.

A solution to this problem might be an API similar to `DefaultSchema` e.g. `DefaultCatalog`. When provided, the transport would use that catalog instead of the one from the connection string. This allows the business code to reach its DB objects in the business DB while SQLT uses objects from `DefaultCatalog`.

**Comment by SzymonPobiega** [comment]
@SimonCropp does it describe the problem/solution well enough?

**Comment by tmasternak** [comment]
Another approach to solving the problem would be to:
 * Use endpoint in `TransactionScope` mode
 * Use endpoint specific connection string per endpoint e.g. "Initial Catalog=Orders"
 * Register custom connection factories for SqlT and SqlP with a callback that uses the default connection string (to enable pooling) but change the database after opening. This will work as the connection pooling invokes `sp_reset_connection` when the connection gets back to the pool and this brings back the initial catalog to the one specified in the connection string. E.g.
 
``` csharp
transport.UseCustomSqlConnectionFactory(
    sqlConnectionFactory: async () =>
    {
        var connection = new SqlConnection("SomeConnectionString");
        await connection.OpenAsync().ConfigureAwait(false);
        connection.ChangeDatabaseAsync("NserviceBus");
    });
```

**Comment by SimonCropp** [comment]
@tmasternak 

> Use endpoint in TransactionScope mode

did u smoke test this approach?

**Comment by SimonCropp** [comment]
@SzymonPobiega 

> does it describe the problem/solution well enough?

yes i think so

**Comment by tmasternak** [comment]
@SimonCropp what I've tested is that the connection which gets returned to the pool has its initial catalog reset before it's reused.

**Comment by SimonCropp** [comment]
@tmasternak so you are saying [this scenario](https://github.com/SimonCropp/nsbdtc) will work with without dtc by using TransactionScope   ?

**Comment by tmasternak** [comment]
@SimonCropp please have a look at https://github.com/SimonCropp/nsbdtc/pull/1.

It's a sample on how to prevent DTC escalation. In the PR business and persistence tables are in dedicated catalogs and the transport tables are in `nservicebus` catalog.

The configuration is a bit cumbersome as there is no `DefaultCatalog` on the transport as @SzymonPobiega already proposed in this issue.

**Comment by SimonCropp** [comment]
@tmasternak i tried that and i still get dtc

![image](https://user-images.githubusercontent.com/122666/107711548-f2347200-6d1b-11eb-944d-7b5bf4be9f97.png)

so that approach works for when u use both a local db and a remote db?

**Comment by tmasternak** [comment]
@SimonCropp the exception that you are getting is coming from a check that the transport does when starting on Core with TransactionScope. This exception results in WARN message saying that TransactionScope+Core might not be a good option if you are not sure you won't get DTC escalation.

If you let the exception bubble and be handled the endpoints should start just fine.

**Comment by SimonCropp** [comment]
@tmasternak yes that seem to work. is there any way that this can be a supported OOTB feature for the sql transport?

**Comment by SimonCropp** [comment]
@tmasternak  actually our persistence is in the NSB DB. will your proposed approach still work?

**Comment by tmasternak** [comment]
> actually our persistence is in the NSB DB. will your proposed approach still work?

no :(. 

The current approach assumes that the catalog is changed from the value in the connection string only in the transport. Currently, there is no way to do that in the persister. It might be possible to tweak the connection used in the handlers but if possible that would require a custom behavior that would change the catalog back and forward, before and after a handler is called.

> is there any way that this can be a supported OOTB feature for the sql transport?

Do you have any specific API in mind? The catalog override for SQL persistence could be added in an enhancement release.

**Comment by SimonCropp** [comment]
> Do you have any specific API in mind

perhaps make the `UseCustomSqlConnectionFactory` customization an supported and consumable API? Given the potential fragility moving forward it would also be nice to have the scenario part of the integration test for the sql transport

> The catalog override for SQL persistence could be added in an enhancement release.

can this be fast tracked? it would be nice to remove the synonym hack we currently have in place

**Comment by tmasternak** [comment]
@SimonCropp I've created https://github.com/Particular/NServiceBus.Persistence.Sql/issues/623 and assigned it as a candidate for the next enhancement release.

**Comment by SimonCropp** [comment]
> We had SQL persistence release that just finished so I think it might take some time for the sqlp issue to considered for the next release.

is there any reason to not do a release now?

**Comment by tmasternak** [comment]
@SimonCropp at this point the vehicle for pushing out enhancements to the components like the one we discuss are enhancement releases.

**Comment by kentdr** [comment]
Our fix for this will be released with version 6.3 of the transport.  You can check out the code changes at PR [853](https://github.com/Particular/NServiceBus.SqlServer/pull/853) (master) and [858](https://github.com/Particular/NServiceBus.SqlServer/pull/858) (6.3)
